src/darwin/Framework/CHIP/CHIPDevice.mm

```diff
@@ -173,6 +202,898 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
+            return nil;
+        }
+        NSMutableArray * array = [[NSMutableArray alloc] init];
+        while ((err = data->Next()) == CHIP_NO_ERROR) {
+            chip::TLV::Tag tag = data->GetTag();
+            id value = ObjectFromTLV(data);
+            if (value == nullptr) {
+                CHIP_LOG_ERROR("Error when decoding TLV container");
+                return nil;
+            }
+            [array addObject:[NSDictionary dictionaryWithObjectsAndKeys:value, kCHIPValueKey,
+                                           [NSNumber numberWithUnsignedLongLong:(unsigned long long) tag.mVal], kCHIPTagKey, nil]];
+        }
+        if (err != CHIP_END_OF_TLV) {
+            CHIP_LOG_ERROR("Error(%s): TLV container decoding failed", chip::ErrorStr(err));
+            return nil;
+        }
+        err = data->ExitContainer(tlvType);
+        if (err != CHIP_NO_ERROR) {
+            CHIP_LOG_ERROR("Error(%s): TLV container exiting failed", chip::ErrorStr(err));
+            return nil;
+        }
+        return [NSDictionary dictionaryWithObjectsAndKeys:typeName, kCHIPTypeKey, array, kCHIPValueKey, nil];
+    }
+    default:
+        CHIP_LOG_ERROR("Error: Unsupported TLV type for conversion: %u", (unsigned) data->GetType());
+        return nil;
+    }
+}
+
+static CHIP_ERROR EncodeTLVFromObject(id object, chip::TLV::TLVWriter & writer, chip::TLV::Tag tag)
+{
+    if (![object isKindOfClass:[NSDictionary class]]) {
+        CHIP_LOG_ERROR("Error: Unsupported object to encode: %@", [object class]);
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    NSString * typeName = ((NSDictionary *) object)[kCHIPTypeKey];
+    id value = ((NSDictionary *) object)[kCHIPValueKey];
+    if (!typeName) {
+        CHIP_LOG_ERROR("Error: Object to encode is corrupt");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if ([typeName isEqualToString:kCHIPSignedIntegerValueTypeKey]) {
+        if (![value isKindOfClass:[NSNumber class]]) {
+            CHIP_LOG_ERROR("Error: Object to encode has corrupt signed integer type: %@", [value class]);
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        return writer.Put(tag, [value intValue]);
+    }
+    if ([typeName isEqualToString:kCHIPUnsignedIntegerValueTypeKey]) {
+        if (![value isKindOfClass:[NSNumber class]]) {
+            CHIP_LOG_ERROR("Error: Object to encode has corrupt unsigned integer type: %@", [value class]);
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        return writer.Put(tag, [value unsignedIntValue]);
+    }
+    if ([typeName isEqualToString:kCHIPBooleanValueTypeKey]) {
+        if (![value isKindOfClass:[NSNumber class]]) {
+            CHIP_LOG_ERROR("Error: Object to encode has corrupt boolean type: %@", [value class]);
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        return writer.Put(tag, [value unsignedIntValue] ? true : false);
+    }
+    if ([typeName isEqualToString:kCHIPFloatValueTypeKey]) {
```

Comment:
I don't think current `CHIPTLVReader` provides a way to distinguish the two.

Reply:
It actually does, though it's not very obvious...  If the type is `kTLVType_FloatingPointNumber` we could try to decode as float, and if that works it's a float.  If not, try to decode as double.

And file a followup to add separate TLVType values for float and double, since they are not interchangeable (unlike various integer sizes).